File: C++/coinChange.cpp
/**
 * @File    : coinChange.cpp
 * @Brief   : 零钱兑换 背包问题
 * @Link    : https://leetcode-cn.com/problems/coin-change/
 * @Author  : Wei Li
 * @Date    : 2021-07-15
*/

#include <iostream>
#include <vector>
#include <algorithm>

class Solution
{
public:
    int coinChange(std::vector<int> &coins, int amount)
    {
        if (coins.empty())
            return -1;

        std::vector<int> dp(amount + 1, amount + 2);
        dp[0] = 0;
        for (int i = 1; i <= amount; ++i)
        {
            for (const int &coin : coins)
            {
                if (i >= coin)
                {
                    dp[i] = std::min(dp[i], dp[i - coin] + 1);
                }
            }
        }
        return dp[amount] == amount + 2 ? -1 : dp[amount];
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    std::vector<int> coins{1, 2, 5};
    int amount = 11;

    Solution solution;
    auto min_coins = solution.coinChange(coins, amount);
    std::cout << "The solution of this problem is " << min_coins << std::endl;

    return 0;
}

File: Python/coinChange.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: coinChange.py
@Function: 零钱兑换 背包问题 - 动态规划
@Link: https://leetcode-cn.com/problems/coin-change/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-15
"""

from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [float('inf')] * (amount + 1)
        dp[0] = 0

        for coin in coins:
            for x in range(coin, amount + 1):
                dp[x] = min(dp[x], dp[x - coin] + 1)
        return dp[amount] if dp[amount] != float('inf') else -1


# -----------------------------
if __name__ == "__main__":
    coins = [1, 2, 5]
    amount = 11

    solution = Solution()
    min_coins = solution.coinChange(coins, amount)
    print(f"The solutio of this problem is {min_coins}")
